packages/x-components/src/x-modules/url/components/url-handler.vue

```diff
@@ -96,31 +122,130 @@
-      this.$x.emit('ExtraParamsLoadedFromUrl', extra, { feature: 'url' });
+      const metadata = this.createWireMetadata();
+      this.$x.emit('ParamsLoadedFromUrl', all, metadata);
+      this.$x.emit('ExtraParamsLoadedFromUrl', extra, metadata);
       // TODO: Move this logic from here.
       if (all.query) {
-        this.$x.emit('UserOpenXProgrammatically');
+        this.$x.emit('UserOpenXProgrammatically', undefined, metadata);
       }
       this.urlLoaded = true;
     }
 
+    /**
+     * Creates the wire metadata to include in every emitted {@link XEvent | XEvents}.
+     *
+     * @returns The {@link WireMetadata | metadata}.
+     * @internal
+     */
+    protected createWireMetadata(): Pick<WireMetadata, 'feature' | 'location'> {
+      return {
+        feature: 'url',
+        location: this.detectLocation()
+      };
+    }
+
+    /**
+     * Detects the {@link FeatureLocation | location} used to build the
+     * {@link QueryOriginInit | events metadata origin}.
+     *
+     * @returns The {@link FeatureLocation | location}.
+     * @internal
+     */
+    protected detectLocation(): FeatureLocation {
+      const currentUrl = new URL(window.location.href);
+      const previousUrl = this.url;
+      this.url = currentUrl;
+
+      const isInternalNavigation =
+        previousUrl?.search !== currentUrl.search && previousUrl?.pathname === currentUrl.pathname;
+      if (isInternalNavigation) {
+        return 'url_history';
+      }
+
+      const isNavigatingFromPDP = this.isNavigatingFromPdp();
+      if (isNavigatingFromPDP) {
+        return 'url_history_pdp';
+      }
+
+      return 'external';
+    }
+
+    /**
+     * Check if the navigation is from a product page.
+     *
+     * @remarks Due to Safari 14 not supporting the new and standard PerformanceNavigationTiming
+     * API, we are falling back to the deprecated one, PerformanceNavigation. We also fallback to
+     * this API whenever we get a navigationType equal to reload, because Safari has a bug that the
+     * navigationType is permanently set to reload after you have reload the page and it never
+     * resets. As some browsers have a back-forward cache implemented, we also take into account if
+     * the page is persisted.
+     *
+     * @returns True if the navigation is from a product page, false otherwise.
+     * @internal
+     */
+    protected isNavigatingFromPdp(): boolean {
+      const isPagePersisted = this.isPagePersisted;
+      const navigationEntries = window.performance.getEntriesByType('navigation');
+      const navigationType = (navigationEntries[0] as PerformanceNavigationTiming)?.type;
+      const useFallbackStrategy =
+        window.performance.navigation &&
+        (isArrayEmpty(navigationEntries) || navigationType === 'reload');
+
+      // Reset internal isPagePersisted property value
+      this.isPagePersisted = false;
+
+      if (useFallbackStrategy) {
```

Comment:
Apparently when you reload a URL with a search, then navigate to the product page, and go back, the timing API still says `reload`, what I'm guessing is that this navigation API is returning the previous type of how you arrived to the search, and is not related with using `back` and `forward` action buttons. But as this is very rare, I think we can leave it as known issue to check it in the future.

https://user-images.githubusercontent.com/68222542/144417743-8f2d1458-23bd-4b20-9098-bec37f8c522b.mov

Reply:
I tried doing the same steps as in your video locally and it was working as expected. I wasn't in `SPA` mode, though. As we discussed, I'll try to take a deeper look at the `pagehide`, `pageshow` and `popstate` events in order to try to replicate this feature without relying in the `performance navigation API` which doesn't seem to be very robust, as we saw the issues it has in the different browsers.

Reply:
Fuck, sorry, I forgot to clarify this in the previous comment, as we were discussing this in private